builtin/providers/nomad/resource_job.go

```diff
@@ -0,0 +1,168 @@
+package nomad
+
+import (
+	"bytes"
+	"encoding/gob"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/nomad/api"
+	"github.com/hashicorp/nomad/jobspec"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceJob() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceJobRegister,
+		Update: resourceJobRegister,
+		Delete: resourceJobDeregister,
+		Read:   resourceJobRead,
+		Exists: resourceJobExists,
+
+		Schema: map[string]*schema.Schema{
+			"jobspec": {
+				Description: "Job specification. If you want to point to a file use the file() function.",
+				Required:    true,
+				Type:        schema.TypeString,
+			},
+
+			"deregister_on_destroy": {
+				Description: "If true, the job will be deregistered on destroy.",
+				Optional:    true,
+				Default:     true,
+				Type:        schema.TypeBool,
+			},
+
+			"deregister_on_id_change": {
+				Description: "If true, the job will be deregistered when the job ID changes.",
+				Optional:    true,
+				Default:     true,
+				Type:        schema.TypeBool,
+			},
+		},
+	}
+}
+
+func resourceJobRegister(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	// Get the jobspec itself
+	jobspecRaw := d.Get("jobspec").(string)
+
+	// Parse it
+	jobspecStruct, err := jobspec.Parse(strings.NewReader(jobspecRaw))
+	if err != nil {
+		return fmt.Errorf("error parsing jobspec: %s", err)
+	}
+
+	// Initialize and validate
+	jobspecStruct.Canonicalize()
+	if err := jobspecStruct.Validate(); err != nil {
+		return fmt.Errorf("Error validating job: %v", err)
+	}
+
+	// If we have an ID and its not equal to this jobspec, then we
+	// have to deregister the old job before we register the new job.
+	prevId := d.Id()
+	if !d.Get("deregister_on_id_change").(bool) {
+		// If we aren't deregistering on ID change, just pretend we
+		// don't have a prior ID.
+		prevId = ""
+	}
+	if prevId != "" && prevId != jobspecStruct.ID {
+		log.Printf(
+			"[INFO] Deregistering %q before registering %q",
+			prevId, jobspecStruct.ID)
+
+		log.Printf("[DEBUG] Deregistering job: %q", prevId)
+		_, _, err := client.Jobs().Deregister(prevId, nil)
+		if err != nil {
+			return fmt.Errorf("error deregistering previous job %q: %s", prevId, err)
```

Comment:
I don't want the user to be confused why we may be deregistering a job "foo" when the jobspec perhaps has job "bar". I've updated the wording to be a bit more explicit about that, pushing soon.